File: src/01.exs
import Enum, only: [count: 2, map: 2, zip: 2]
import IO, only: [puts: 1, read: 2]
import String, only: [split: 3, to_integer: 1]

defmodule Day01 do
  @huge 9999

  ##
  # The "solve_1" solution is more in line with most other languages solutions
  def solve_1(numbers) do
    solve_1(numbers, @huge, 0)
  end

  def solve_1([], _, result) do
    result
  end

  def solve_1([n|rest], previous, result) do
    case n > previous do
      true -> solve_1(rest, n, result + 1)
      false -> solve_1(rest, n, result)
    end
  end

  ##
  # Here, the "part_1" solutions are more in line with the Haskell solutions
  def part_1(numbers) do
    zip(numbers, tl(numbers))
    |> count(fn {left, right} -> left < right end)
  end

  def part_2(numbers) do
    sum_by_3(numbers)
    |> part_1
  end

  defp sum_by_3(numbers) do
    case numbers do
      [x, y, z | rest] -> [x+y+z] ++ sum_by_3([y, z | rest])
      _ -> []
    end
  end
end

numbers = read(:stdio, :all)
          |> split("\n", trim: true)
          |> map(&to_integer/1)

puts Day01.solve_1(numbers)
puts Day01.part_2(numbers)

File: src/01.erl
-module(aoc).
-export([main/1]).
-include("common/common.hrl").

-define(HUGE, 9999).

solve_1(Numbers) ->
    solve_1(Numbers, ?HUGE, 0).

solve_1([], _, Result) ->
    Result;

solve_1([N | Rest], Previous, Result) ->
    case N > Previous of
        true -> solve_1(Rest, N, Result + 1);
        false -> solve_1(Rest, N, Result)
    end.

solve_2(Numbers) ->
    solve_2(Numbers, ?HUGE, ?HUGE, ?HUGE, 0).

solve_2([], _, _, _, Result) ->
    Result;

solve_2([N | Rest], P1, P2, P3, Result) ->
    case (N+P1+P2 > P1+P2+P3) of
        true -> solve_2(Rest, N, P1, P2, Result + 1);
        false -> solve_2(Rest, N, P1, P2, Result)
    end.

main([Filename]) ->
    Numbers = get_parsed_lines(Filename, fun erlang:list_to_integer/1),
    io:fwrite("~w~n", [solve_1(Numbers)]),
    io:fwrite("~w~n", [solve_2(Numbers)]).
